### ISSUE TITLE
How mqtt connection is distributed to clustered(on k8s) emqx nodes?

### ISSUE THREAD

**Issue by gomgomshrimp** [emqx/emqx#4467]
I'm running emqx cluster on k8s and I checked mqtt connection is distributed to nodes. 

I read that emqx cluster is usually deployed behind LB... but I'm not built any LB for emqx cluster on k8s.  

Then how mqtt connection is distributed to nodes? Is there its own LB when building emqx cluster on k8s?

**Comment by HJianBo** [comment]
It depends only on which node the client is connected to, and on the front Proxy's strategy for distributing TCP connections.

**Comment by gomgomshrimp** [comment]
@HJianBo Thanks to reply.
But there are some misunderstood for my question. I'm not asking about connection from external to internal...
 
I already built the emqx cluster on EKS and it works well. There are three emqx nodes for clustering, and mqtt connections are distributed well, but I didn't set any Load Balancer. (Each emqx node uses "headless service" - kubernetes service that has no load balancing)

Please check emqx cluster by using [autocluster based k8s](https://docs.emqx.io/en/broker/v4.3/advanced/cluster.html#node-discovery-and-automatic-clustering). Maybe Kubernetes does it automatically not emqx....

**Comment by HJianBo** [comment]
Sorry @gomgomshrimp I don't know much about EKS, but I'm sure there must be services in the middle that exist for LB functions. Because EMQ X nodes can't do connection distribution by themselves. We formed the cluster by configuring K8S just to find the node names of other emqx nodes.

**Comment by bloodberrys** [comment]
@gomgomshrimp Well, I have the same question to this issue. Did you have any solution already? Anyway, I deployed the emqx using helm charts based on the emqx broker doc here: https://docs.emqx.io/en/broker/v4.3/getting-started/install.html#install-emq-x-in-docker-contain-a-simple-docker-compose-cluster to my EKS, but there was no loadBalancer available on the emqx-service. All tutorial available for fully-managed kubernetes (such as kubeadm, minikube, etc), but not particularly provided for EKS. Then, how to expose this service in EKS?

**Comment by bloodberrys** [comment]
I have tried to patch the service to 
`kubectl patch svc my-emqx -p '{"spec": {"type": "LoadBalancer"}}'`
and also edit the service to add an NLB loadbalancer annotations for AWS. The external IP showed up but I still can't pub/sub with the DNS.

**Comment by zmstone** [comment]
If I remember correctly, service type can not be changed without redeployment.

**Comment by zmstone** [comment]
> @HJianBo Thanks to reply.
> But there are some misunderstood for my question. I'm not asking about connection from external to internal...
> 
> I already built the emqx cluster on EKS and it works well. There are three emqx nodes for clustering, and mqtt connections are distributed well, but I didn't set any Load Balancer. (Each emqx node uses "headless service" - kubernetes service that has no load balancing)
> 
> Please check emqx cluster by using [autocluster based k8s](https://docs.emqx.io/en/broker/v4.3/advanced/cluster.html#node-discovery-and-automatic-clustering). Maybe Kubernetes does it automatically not emqx....

Hi. I am not sure if I understand the questions correctly but
1.  EQM X nodes discover each other to form a cluster using k8s APIs.
2. K8s Service publishing (so to reachable from external)
These are two different topics, I don't think they are related.

I think @bloodberrys is going in the right direction (by adding a load balancer).
ref: https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types

**Comment by bloodberrys** [comment]
@zmstone exactly. 
This is my attempt:
0. Adding tcp port 1883 to my AWS security group.
1. `kubectl edit svc/my-emqx` and adding `service.beta.kubernetes.io/aws-load-balancer-type: "nlb"` under annotations.
2. exit
3. kubectl patch svc my-emqx -p '{"spec": {"type": "LoadBalancer"}}'
4. I got my external IP!!
![image](https://user-images.githubusercontent.com/36856722/114298390-5edfa900-9ae0-11eb-9ead-d9edc6c8bf0b.png)
5. `nslookup` my external IP:
![image](https://user-images.githubusercontent.com/36856722/114298403-7159e280-9ae0-11eb-91fb-e7662432cae9.png)
seems great.

but when I Tried to pub/sub:
![image](https://user-images.githubusercontent.com/36856722/114298441-a36b4480-9ae0-11eb-81c5-a917d4602284.png)
I got `All subscription requests were denied.`. 

What does `All subscription requests were denied.` stand for?
